src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,766 @@
+    {
+        if (FutureArrays.compareUnsigned(packedValue.bytes,
+                                         packedValue.offset,
+                                         packedValue.offset + bytesPerValue,
+                                         minPackedValue,
+                                         0,
+                                         bytesPerValue) < 0)
+        {
+            throw new AssertionError("value=" + new BytesRef(packedValue.bytes, packedValue.offset, bytesPerValue) +
+                                     " is < minPackedValue=" + new BytesRef(minPackedValue));
+        }
+
+        if (FutureArrays.compareUnsigned(packedValue.bytes,
+                                         packedValue.offset,
+                                         packedValue.offset + bytesPerValue,
+                                         maxPackedValue, 0,
+                                         bytesPerValue) > 0)
+        {
+            throw new AssertionError("value=" + new BytesRef(packedValue.bytes, packedValue.offset, bytesPerValue) +
+                                     " is > maxPackedValue=" + new BytesRef(maxPackedValue));
+        }
+    }
+
+    private void valuesInOrderAndBounds(int count, byte[] minPackedValue, byte[] maxPackedValue, IntFunction<BytesRef> values, long[] docs)
+    {
+        byte[] lastPackedValue = new byte[bytesPerValue];
+        long lastDoc = -1;
+        for (int i = 0; i < count; i++)
+        {
+            BytesRef packedValue = values.apply(i);
+            assert packedValue.length == bytesPerValue;
+            valueInOrder(i, lastPackedValue, packedValue.bytes, packedValue.offset, docs[i], lastDoc);
+            lastDoc = docs[i];
+
+            // Make sure this value does in fact fall within this leaf cell:
+            valueInBounds(packedValue, minPackedValue, maxPackedValue);
+        }
+    }
+
+    private void valueInOrder(long ord, byte[] lastPackedValue, byte[] packedValue, int packedValueOffset, long doc, long lastDoc)
+    {
+        int dimOffset = 0;
+        if (ord > 0)
+        {
+            int cmp = FutureArrays.compareUnsigned(lastPackedValue, dimOffset, dimOffset + bytesPerValue, packedValue, packedValueOffset + dimOffset, packedValueOffset + dimOffset + bytesPerValue);
+            if (cmp > 0)
+            {
+                throw new AssertionError("values out of order: last value=" + new BytesRef(lastPackedValue) +
+                                         " current value=" + new BytesRef(packedValue, packedValueOffset, bytesPerValue) +
+                                         " ord=" + ord);
+            }
+            if (cmp == 0 && doc < lastDoc)
+            {
+                throw new AssertionError("docs out of order: last doc=" + lastDoc + " current doc=" + doc + " ord=" + ord);
+            }
+        }
+        System.arraycopy(packedValue, packedValueOffset, lastPackedValue, 0, bytesPerValue);
+    }
+}
```

Comment:
nit: Given `OneDimensionBKDWriter` isn't static, should we just move the methods here into it and make/keep them private if possible? Not sure why some are in there and others aren't...

Reply:
Agreed. I have also done a little reorganisation of the whole class so that it's easier to read. I've moved the inner interface and classes to the bottom of the class with the `TreeWriter` owning its own methods. I've changed the behaviour of the `TreeWriter.finish` method so it now returns the total point count with the `BlockBalancedTreeWriter.writeTree` method assuming responsibility for fetching the file pointer.